### PR TITLE
Add marshaling between Dynamo and Python dictionaries

### PR DESCRIPTION
**Pull request by aparajit-pratap**
### Purpose

Add Dynamo Dictionary <-> Python Dictionary marshaling support.
JIRA: https://jira.autodesk.com/browse/QNTM-2990

image

### Declarations

Check these if you believe they are true

- [x] The code base is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [x] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [x] All tests pass using the self-service CI.
- [x] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions), and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.

I've yet to add tests and run self-service - WIP

### Reviewers

@alfarok 
@mjkkirschner

### CODE REVIEW

File: src/DynamoUtilities/DataMarshaler.cs [hunk 1]
```diff
@@ -18,11 +18,13 @@ public class DataMarshaler
...
         public DataMarshaler()
         {
-            //RegisterMarshaler((IEnumerable e) => e.Cast<object>().Select(Marshal));
             RegisterMarshaler((IList e) => e.Cast<object>().Select(Marshal).ToList());
             RegisterMarshaler(
                 (IDictionary dict) =>
-                    dict.Cast<dynamic>().ToDictionary(x => Marshal(x.Key), x => Marshal(x.Value)));
+                {
+                    // Dictionary<TKey, TValue> and IronPython.Runtime.PythonDictionary both implement IDictionary
+                    return dict.Keys.Cast<object>().ToDictionary(key => Marshal(key), key => Marshal(dict[key]));
```
Comment (mjkkirschner) [inline src/DynamoUtilities/DataMarshaler.cs]: does this work nicely for nested dictionaries?

**Comment by mjkkirschner** [comment]
@aparajit-pratap looks good can we add some tests? Specifically I'm curious about nested dictionaries as values?

**Comment by aparajit-pratap** [comment]
@mjkkirschner seems to work for lists and nested dictionaries.

**Comment by aparajit-pratap** [comment]
@alfarok  I'm unable to open the DYN files you created for JSON tests.

**Comment by mjkkirschner** [comment]
LGTM

**Pull request merged**